### ISSUE TITLE
Chat History persists with `ollama run`

### ISSUE THREAD

**Issue by hg0428** [ollama/ollama#7152]
### What is the issue?

![image](https://github.com/user-attachments/assets/b53c625b-0448-4d10-aa0b-63945f72c02e)
See the image. The chat history is persisting for [Falcon](https://ollama.com/Hudson/falcon-mamba-instruct), but not any other models (at least that I've tried so far). I would expect it to not persist. It even persists after re-creating the model:
![image](https://github.com/user-attachments/assets/03cbe28e-9fbd-452a-b53b-503ec3c67d50)

### OS

macOS

### GPU

Apple

### CPU

Apple

### Ollama version

0.3.12

**Comment by hg0428** [comment]
The problem happens in the API as well.
![image](https://github.com/user-attachments/assets/1c38ffc5-a7be-4cbd-a3b8-cfa323cc2b44)

**Comment by rick-github** [comment]
What's in your `Modelfile`?  What's the output of `ollama show --modelfile falcon-mamba-instruct:7b-q4_0`?

**Comment by hg0428** [comment]
> What's in your `Modelfile`? What's the output of `ollama show --modelfile falcon-mamba-instruct:7b-q4_0`?

```
# Modelfile generated by "ollama show"
# To build a new Modelfile based on this, replace FROM with:
# FROM falcon-mamba-instruct:7b-q4_0

FROM /Users/hudsongouge/.ollama/models/blobs/sha256-26e1d318e3c108d4f263056618afc3bf181fd518011451f3ea814eef03627157
TEMPLATE "{{- range .Messages }}<|im_start|>{{ .Role }}
{{ .Content }}<|im_end|>
{{ end }}<|im_start|>assistant
"
SYSTEM You are a helpful AI assistant.
PARAMETER stop <|im_start|>
PARAMETER stop <|im_end|>
PARAMETER min_p 0.1
PARAMETER repeat_penalty 1
```

**Comment by victorb** [comment]
To me this seems like a bug in the Python library you're using for Ollama, not in Ollama itself, or the API.

My own example:

```rust
    println!("Step #1");

    let messages = vec![
        client::Message::new("system", r#"You are a helpful assistant"#),
        client::Message::new("user", "x = 4 and y = 137. Remember that."),
    ];

    let response: LLMResponse = client::call_llm(&client, model, messages.clone())
        .await
        .unwrap();

    println!("Response:\n{:#?}", response.message.content);

    println!("Step #2");

    let messages = vec![
        client::Message::new("system", r#"You are a helpful assistant"#),
        client::Message::new("user", "What are x and y?"),
    ];

    let response: LLMResponse = client::call_llm(&client, model, messages.clone())
        .await
        .unwrap();

    println!("Response:\n{:#?}", response.message.content);
```

Outputs:

```
Step #1
Response:
"I've noted that you've set x to 4 and y to 137. If you need any calculations or further information related to these values, feel free to ask!"
Step #2
Response:
"To provide an answer, I need more information about what x and y represent. Could you please give me the context or equations involving x and y?"
```

Which seems to work like I expect it to.

Can you replicate this issue when you use the HTTP API directly without using the Python library you're using?

**Comment by hg0428** [comment]
> To me this seems like a bug in the Python library you're using for Ollama, not in Ollama itself, or the API.
> 
> My own example:
> 
> ```rust
>     println!("Step #1");
> 
>     let messages = vec![
>         client::Message::new("system", r#"You are a helpful assistant"#),
>         client::Message::new("user", "x = 4 and y = 137. Remember that."),
>     ];
> 
>     let response: LLMResponse = client::call_llm(&client, model, messages.clone())
>         .await
>         .unwrap();
> 
>     println!("Response:\n{:#?}", response.message.content);
> 
>     println!("Step #2");
> 
>     let messages = vec![
>         client::Message::new("system", r#"You are a helpful assistant"#),
>         client::Message::new("user", "What are x and y?"),
>     ];
> 
>     let response: LLMResponse = client::call_llm(&client, model, messages.clone())
>         .await
>         .unwrap();
> 
>     println!("Response:\n{:#?}", response.message.content);
> ```
> 
> Outputs:
> 
> ```
> Step #1
> Response:
> "I've noted that you've set x to 4 and y to 137. If you need any calculations or further information related to these values, feel free to ask!"
> Step #2
> Response:
> "To provide an answer, I need more information about what x and y represent. Could you please give me the context or equations involving x and y?"
> ```
> 
> Which seems to work like I expect it to.
> 
> Can you replicate this issue when you use the HTTP API directly without using the Python library you're using?

The issue is also in the Ollama cli via `ollama run <model>`, so I would expect it to be an issue with Ollama. The Python library I used was the official ollama python library.

**Comment by victorb** [comment]
> The issue is also in the Ollama cli via ollama run <model>, so I would expect it to be an issue with Ollama. The Python library I used was the official ollama python library.

I can't seem to reproduce this issue with `ollama run` either, works like I expect it to too:

```bash
$ ./ollama run qwen2.5:32b "x = 4 and y = 137. Remember that."
Got it, you've specified that x = 4 and y = 137. If you need any calculations or further information involving these variables,
feel free to ask!

$ ./ollama run qwen2.5:32b "What are x and y?"
To provide an answer about what \(x\) and \(y\) are, I would need more context or information regarding the specific scenario
you're referring to, as \(x\) and \(y\) can represent different things depending on the situation:

1. **Mathematics**: In mathematics, \(x\)^C
```

**Comment by hg0428** [comment]
> > The issue is also in the Ollama cli via ollama run , so I would expect it to be an issue with Ollama. The Python library I used was the official ollama python library.
> 
> I can't seem to reproduce this issue with `ollama run` either, works like I expect it to too:
> 
> ```shell
> $ ./ollama run qwen2.5:32b "x = 4 and y = 137. Remember that."
> Got it, you've specified that x = 4 and y = 137. If you need any calculations or further information involving these variables,
> feel free to ask!
> 
> $ ./ollama run qwen2.5:32b "What are x and y?"
> To provide an answer about what \(x\) and \(y\) are, I would need more context or information regarding the specific scenario
> you're referring to, as \(x\) and \(y\) can represent different things depending on the situation:
> 
> 1. **Mathematics**: In mathematics, \(x\)^C
> ```

If you see my original message, the issue ONLY occurs with Falcon Mamba. That's the weird part. Either way, it is a problem that needs to be fixed as that model is pretty much unusable for my user-case until this is fixed.

**Comment by hg0428** [comment]
![image](https://github.com/user-attachments/assets/5ebd3327-a0d5-4922-9062-b5824aeef963)

**Comment by YonTracks** [comment]
I see something happening with the run cmd?  ``` runCmd := &cobra.Command{
        Use:     "run MODEL [PROMPT]",
        Short:   "Run a model",
        Args:    cobra.MinimumNArgs(1),
        PreRunE: checkServerHeartbeat,
        RunE:    RunHandler,
    }```  if I check the server.log I can see the chat endpoint being called?  messages / context / cache / persistence  issue there somewhere? I keep looking good luck.

**Comment by YonTracks** [comment]
I see, if I use `ollama run llama3.1 what can you do? ` it uses the generate (love? seems more smart? powerful). but yes, if only ` ollama run llama3.1` then it uses the chat with persisted messages / prompts (issue). good luck. I will still try see what I can find (chat and generate in a single function seems best) anyway good luck.

**Comment by pdevine** [comment]
My guess here is the problem is with falcon-mamba and the kv cache. Llama3.1 and llama3.2 seem to be working correctly.

**Comment by hg0428** [comment]
> My guess here is the problem is with falcon-mamba and the kv cache. Llama3.1 and llama3.2 seem to be working correctly.

Yes, all other models I have tested (llama, Owen, gemma, etc...) work correctly.

**Comment by jessegross** [comment]
Yes, recurrent models such as Mamba do use the KV cache differently, so that is likely where the problem is.

**Comment by jessegross** [comment]
I don't see this with the most recent 0.4.0-rc, can you see if it is still a problem there?
